API Developers Guide (#4053)
circleci/circleci-docs: jekyll/_cci2/api-developers-guide.md

---
layout: classic-docs
title: "CircleCI API Developer's Guide"
short-title: "Developer's Guide"
description: "API cookbook for internal and external CircleCI developers"
categories: [getting-started]
order: 1
---

This *API Developer's Guide* was written to assist developers in quickly and easily making API calls to CircleCI services to return detailed information about users, pipelines, projects and workflows.

* TOC
{:toc}

# API Overview

The CircleCI platform provides a powerful API that enables users to retrieve detailed information about users, jobs, workflows and pipelines.

CircleCI API v2 enables you to use endpoints with several new features that improve the API experience, in addition to optimizing how you use the API for your jobs. Please note that CircleCI API v2 is currently in active development, therefore, the stability of the API is referred to as “mixed”.

> ## API Classes

Reviewer: I suggest you don't try to keep this in sync -- let the generated docs speak for themselves on this.

Author: Removed from doc.

The current classes of the API v2 endpoints are:

- Authentication
- Pipeline
- Workflows
- User (Preview)
- Project (Preview)
- Job (Preview)

> **Note:** Portions of the CircleCI API v2 remain under “Preview”. Preview endpoints are not yet fully supported or considered generally available. Breaking changes to API v2 Preview endpoints are planned in advance and are announced in the API v2 breaking changes log.

Author: @teesloane I am not sure. I will reach out to @ndintenfass and @jama22 to find out if there will be a changelog for API v2.

> **Note:** Portions of the CircleCI API v2 remain under “Preview”. Preview endpoints are not yet fully supported or considered generally available. Breaking changes to API v2 Preview endpoints are planned in advance and are announced in the API v2 breaking changes log.

Author: There is a "breaking changes" doc that was created for the initial API v2 rollout, but it is not something that I think we should link to.

https://github.com/CircleCI-Public/api-preview-docs/blob/master/docs/breaking.md

> **Note:** Portions of the CircleCI API v2 remain under “Preview”. Preview endpoints are not yet fully supported or considered generally available. Breaking changes to API v2 Preview endpoints are planned in advance and are announced in the API v2 breaking changes log.

Reviewer 2: Just noticed this breaking changes log is already linked to in the api-intro doc that is live already in the reference section

## Authentication and Authorization

The CircleCI API utilizes token-based authentication to manage access to the API server and validate that a user has permission to make API requests. Before you can make an API request, you must first add an API token and then verify that you are authenticated by the API server to make requests. The process to add an API token and have the API server authenticate you is described in the sections below.

**Note** You may use the API token as the username for HTTP Basic Authentication, by passing the `-u` flag to the `curl` command.

### Add an API Token

To add an API token, perform the steps listed below.

1.  Add an API token from your [account dashboard](https://circleci.com/account/api).
2.  To test it, [View it in your browser](https://circleci.com/api/v1.1/me) or call the API using the command below.

```sh

> $ curl https://circleci.com/api/v1.1/me?circle-token=:token

Author: Added this note to the document.

```

> 3.  You should see a response similar to the example shown below.

Author: Added a note to this section to address this issue.

With API v2, CircleCI is introducing a string representation of the triplet called the `project_slug`, which takes the following form:

`<project_type>/<org_name>/<repo_name>`

The `project_slug` is included in the payload when pulling information about a project as well as when looking up a pipeline or workflow by ID. The `project_slug` can then be used to get information about the project. It is possible in the future the shape of a project_slug may change, but in all cases it would be usable as a human-readable identifier for a given project.

## Rate Limiting/Throttling

> CircleCI does not currently perform any throttling or rate limiting. In the future, CircleCI may institute rate limiting or throttling or both to monitor the number of API calls made to the service and ensure that the system is more stable and has higher overall performance

Reviewer: I don't think we should say this. Invites abuse.

Author: Removed from document

200 - Success
400 - Client error
500 - Server error

## 200 Status Codes

If you receive a 200 HTTP status code, your API request is successful and the requested resource will be returned. The following 200 HTTP status codes could potentially be returned with your request:

`200 - OK`

## 400 Status Codes

If you receive a 400 HTTP status code, there is a problem with the request and the server is unable to successfully process the request. The following status codes may be returned with your request:

`401 - Unauthorized`
`403 - Forbidden`
`404 - Not Found`

## 500 Status Code

If you receive a 500 HTTP status code, there is a problem with the server and the request cannot be processed. If you encounter a 500 response, the error will be logged and CircleCI will work to resolve the error. The following 500 HTTP status codes could potentially be returned with your request:

`500 - Internal Server Error`

For more detailed information about HTTP status codes, refer to the following resource:

> http://www.restapitutorial.com/httpstatuscodes.html

Author: Fixed

---
layout: classic-docs
title: "CircleCI API Developer's Guide"
short-title: "Developer's Guide"
description: "API cookbook for internal and external CircleCI developers"
categories: [getting-started]
order: 1
---

This *API Developer's Guide* was written to assist developers in quickly and easily making API calls to CircleCI services to return detailed information about users, pipelines, projects and workflows.

* TOC
{:toc}

# API Overview

The CircleCI platform provides a powerful API that enables users to retrieve detailed information about users, jobs, workflows and pipelines.

**Note:** Whenever you see a term prefixed with a colon (:), please note that this value must be replaced by a user-input value. For example `:token` in the above call must be replaced by the token string generated in your account:

> curl https://circleci.com/api/v1.1/me?circle-token=dkanmbowk34iajdkl3iakdthisisarandomtoken

Reviewer 2: Replace with example using v2 api?

Reviewer 2: esp here because there doesn't seem to actually be a `:token` part to this request? Maybe that's a v2 thing?

Author: Fixed

- Workflows
- User (Preview)
- Project (Preview)
- Job (Preview)

**Note:** Portions of the CircleCI API v2 remain under “Preview”. Preview endpoints are not yet fully supported or considered generally available. Breaking changes to API v2 Preview endpoints are planned in advance and are announced in the API v2 breaking changes log.

## Authentication and Authorization

The CircleCI API utilizes token-based authentication to manage access to the API server and validate that a user has permission to make API requests. Before you can make an API request, you must first add an API token and then verify that you are authenticated by the API server to make requests. The process to add an API token and have the API server authenticate you is described in the sections below.

**Note** You may use the API token as the username for HTTP Basic Authentication, by passing the `-u` flag to the `curl` command.

### Add an API Token

To add an API token, perform the steps listed below.

1.  Add an API token from your [account dashboard](https://circleci.com/account/api).
2.  To test it, [view it in your browser](https://circleci.com/api/v1.1/me) or call the API using the command below.

```sh

> $ curl https://circleci.com/api/v1.1/me?circle-token=:token

Reviewer 2: As above, maybe we can replace this with a v2 example? - just because that's the focus of the doc

Author: Fixed

  "login" : "someuser"
 }
```

**Note:** All API calls are made in the same way, by making standard HTTP calls, using JSON, a content-type, and your API token. Please note that the JSON examples shown in this document are not comprehensive and may contain additional JSON response fields not shown in the example, based on user input and fields.

### Get Authenticated

To be authenticated by the API server, add an API token using your [account dashboard](https://circleci.com/account/api). To use the API token, add it to the `circle-token` query param:

```sh
curl "https://circleci.com/api/v1.1/me?circle-token=:token"
```

Alternatively, you can use the API token as the username for HTTP Basic Authentication, by passing the `-u` flag to the `curl` command:

```sh 
curl -u <circle-token>: "https://circleci.com/api/..."
```

**Note:** The colon ":" notifies `curl` that there is no password being passed.

## API Endpoints

The CircleCI v2 API, and its associated endpoints allow you to make HTTP calls to designated endpoints developed in the underlying CircleCI API architecture. These endpoints provide programmatic access to CircleCI services, including pipelines, workflows, and jobs. 

> Before working with the API and making calls, you should first have an undestanding of the endpoints currently available. The table below lists the endpoints currently available. 

Reviewer 2: ```suggestion
Before working with the API and making calls, you should first have an understanding of the endpoints that are currently available, as detailed in the table below:
```

`GET /project/:project_slug`  | This endpoint enables users to retrieve a specific Project by its unique slug.
`POST /project/:project_slug/pipeline` | This endpoint enables users to retrieve an individual project by its unique slug.
`GET /pipeline/:id` | This endpoint enables users to retrieve an individual pipeline, based on the `id` passed in the request.
`GET /pipeline/:id/config`  | This endpoint enables users to retrieve the configuration of a specific pipeline.
`GET /project/:project_slug/pipelines/[:filter]`  | This endpoint enables users to retrieve the most recent set of pipelines for a Project.

## API Syntax

When making an API request, make sure you follow standard REST API syntax and formatting. Adhering to proper REST API syntax ensures that the API server can properly process your request and return a valid JSON response. To make a request to the CircleCI API, use the following format:

`https://circleci.com/api/v2`

Where:

- `https://circleci.com` is the resource URL for the API being called.
- `api` is the class being called.
- `v2` is the API version.

# Getting Started with the API

> The CircleCI API v2 is backwards-compatible with previous API versions in the way it identifies your projects using repository name. For instance, if you want to pull information from CircleCI about the GitHub repository https://github.com/CircleCI-Public/circleci-cli you can refer to that in the CircleCI API as gh/CircleCI-Public/circleci-cli, which is a “triplet” of the project type, the name of your “organization”, and the name of the repository. For the project type you can use `  github` or `bitbucket` as well as the shorter forms `gh` or `bb`, which is supported in API v2. The `organization` is your username or organization name in your version control system.

Reviewer 2: ```suggestion
The CircleCI API v2 is backwards-compatible with previous API versions in the way it identifies your projects using repository name. For instance, if you want to pull information from CircleCI about the GitHub repository https://github.com/CircleCI-Public/circleci-cli you can refer to that in the CircleCI API as gh/CircleCI-Public/circleci-cli, which is a “triplet” of the project type, the name of your “organization”, and the name of the repository. For the project type you can use ` github` or `bitbucket` as well as the shorter forms `gh` or `bb`, which are supported in API v2. The `organization` is your username or organization name in your version control system.
```

Reviewer 2: just a fix for the literal formatting that had got a little messed up due to a space - can't actually tell where that is looking at this change request though!

With API v2, CircleCI is introducing a string representation of the triplet called the `project_slug`, which takes the following form:

`<project_type>/<org_name>/<repo_name>`

The `project_slug` is included in the payload when pulling information about a project as well as when looking up a pipeline or workflow by ID. The `project_slug` can then be used to get information about the project. It is possible in the future the shape of a project_slug may change, but in all cases it would be usable as a human-readable identifier for a given project.

# HTTP Status Codes and Response Handling

When you make an API request to the server, an HTTP status code is returned along with the JSON body response. The CircleCI API v2 adheres to standard HTTP response codes, which include the following status code definitions:

200 - Success
400 - Client error

> 500 - Server error

Reviewer 2: This needs extra formatting - e.g. bullets - to make them appear on separate lines

Author: fixed

## 200 Status Codes

If you receive a 200 HTTP status code, your API request is successful and the requested resource will be returned. The following 200 HTTP status codes could potentially be returned with your request:

`200 - OK`

## 400 Status Codes

If you receive a 400 HTTP status code, there is a problem with the request and the server is unable to successfully process the request. The following status codes may be returned with your request:

`401 - Unauthorized`
`403 - Forbidden`

> `404 - Not Found`

Reviewer 2: same as above RE formatting

    ```sh
    curl -X POST https://circleci.com/api/v2/project/gh/{YOUR_USER_NAME}/hello-world/pipeline \
      -H 'Content-Type: application/json' \
      -H 'Accept: application/json' \
      -H "Circle-Token: $CIRCLECI_TOKEN" \
      
    # Which returns:
    {
      "number": 2,
      "state": "pending",
      "id": "e411ea74-c64a-4d60-9292-115e782802ed",
      "created_at": "2020-01-15T15:32:36.605Z"
    }
    ```

    While this alone can be useful, we want to be able to customize parameters of the pipeline when we send this `POST` request. By including a body parameter in the `curl` request (via the `-d` flag), we can customize specific attributes of the pipeline when it runs: pipeline parameters, the branch, or the git tag. Below, we are telling the pipelines to trigger for "my-branch"

    ```sh
    curl -X POST https://circleci.com/api/v2/project/gh/{YOUR_USER_NAME}/hello-world/pipeline \
      -H 'Content-Type: application/json' \
      -H 'Accept: application/json' \
      -H "Circle-Token: $CIRCLECI_TOKEN" \
      -H 'x-attribution-login: string' \
      -H 'x-attribution-actor-id: string' 
      -d "branch=my-branch"

>     ```

Reviewer 2: <img width="1161" alt="Screenshot 2020-01-20 at 11 56 39" src="https://user-images.githubusercontent.com/24589403/72724711-025f3980-3b7c-11ea-8572-6f7f7d490fed.png">

when I try this the pipeline is triggered and then the `-d "branch=my-branch"` part is there ready to be run on it's own, rather than being part of the API post thingy - see screenshot - hope this makes sense!

```sh
$ curl https://circleci.com/api/v2/me
```

3.  You should see a JSON response similar to the example shown below.

```json
{
  "user_key_fingerprint" : null,
  "days_left_in_trial" : -238,
  "plan" : "p16",
  "trial_end" : "2011-12-28T22:02:15Z",
  "basic_email_prefs" : "smart",
  "admin" : true,
  "login" : "someuser"
 }
```

**Note:** All API calls are made in the same way, by making standard HTTP calls, using JSON, a content-type, and your API token. Please note that the JSON examples shown in this document are not comprehensive and may contain additional JSON response fields not shown in the example, based on user input and fields.

### Get Authenticated

To be authenticated by the API server, add an API token using your [account dashboard](https://circleci.com/account/api). To use the API token, add it to the `circle-token` query param:

```sh
curl "https://circleci.com/api/v1.1/me?circle-token=:token"
```

Alternatively, you can use the API token as the username for HTTP Basic Authentication, by passing the `-u` flag to the `curl` command:

```sh 
curl -u <circle-token>: "https://circleci.com/api/..."
```

> **Note:** The colon ":" notifies `curl` that there is no password being passed.

Reviewer 3: I think we generally prefer the header method, so I'd swap these two.

Author: fixed

## API Endpoints

The CircleCI v2 API, and its associated endpoints allow you to make HTTP calls to designated endpoints developed in the underlying CircleCI API architecture. These endpoints provide programmatic access to CircleCI services, including pipelines, workflows, and jobs. 

Before working with the API and making calls, you should first have an understanding of the endpoints that are currently available, as detailed in the tables below:

### User Endpoints

Endpoint       | Description                       
-----------|-------------------------------------------------------

> `/user/{id} ` | This endpoint enables you to return information about a specific user by making a GET request and passing the `id` parameter in the request

Reviewer 3: To follow common conventions I'd add the method to the first column, like `GET /user/{id}` everywhere, instead of having it somewhere in the description.

I'd also shorten the description by a lot in general, from

```
This endpoint enables you to return information about a specific user by making a GET request and passing the `id` parameter in the request
```

to something like

```
Returns information about a specific user
```

Author: Fixed in all tables.

`/project/{project-slug}/{job-number}/artifacts`  | This endpoint enables users to retrieve artifacts for a job by making a GET request and passing the project-slug and job-number parameters in the request.
`/project/{project-slug}/{job-number}/tests` | This endpoint enables you to retrieve test metadata for a job by making a GET request and passing the `project-slug` and `job-number` parameters in the request.

### Workflow Endpoints

Endpoint       | Description                       
-----------|-------------------------------------------------------
`/workflow/{id}` | This endpoint enables users to you to return a workflow by making a GET request and passing the `id` parameter in the request.
`/workflow/{id}/cancel` | This endoint enables you to cancel a workflow by making a POST request and passing the `id` parameter in the request.
`/workflow/{id}/job`  | This endpoint enables you to return the jobs associated with a workflow by making a GET request and passing the `id` parameter in the request.
`/workflow/{id}/rerun` | This endpoint enables you to rerun a workflow by making a POST request and passing the `id` parameter in the request.

> ## API Syntax

Reviewer 3: Not sure this section is actually interesting to the user.

Author: commented out on page.

When making an API request, make sure you follow standard REST API syntax and formatting. Adhering to proper REST API syntax ensures the API server can properly process your request and return a valid JSON response. To make a request to the CircleCI API, use the following format:

`https://circleci.com/api/v2`

Where:

- `https://circleci.com` is the resource URL for the API being called.
- `api` is the service being called.
- `v2` is the API version.

> # Getting Started with the API

Reviewer 3: This is a stylistic choice, but I'd pull the "getting started" part before the huge list of available endpoints, which imho should probably go at the very end as it's more a reference than an explanation or narrative.

Author: Agreed.....went ahead and moved the list of endpoints to the end of the page in a Reference section.

## Prerequisites

* A GitHub or BitBucket account with a repository to setup with CircleCI.
* Completion of the CircleCI onboarding.
* **Optional**: You can format JSON responses by piping the `curl` command into the `jq` utility if you have it installed: `curl ... | jq`.

## Steps

1. On your VCS provider, create a repository. The repo for this example will be called `hello-world`.

2. Next, follow the onboarding for a new project on CircleCI. You can access onboarding by visiting the application and clicking on "Add Projects" in the sidebar or by going to the link: https://onboarding.circleci.com/project-dashboard/{VCS}/{ORG_NAME} where `VCS` is either `github` (or `gh`) or `bitbucket` (or `bb`) and `ORG_NAME` is your organization or personal VCS username. Find your project in the onboarding list and click `Setup Project`. After completing an onboarding, you should have a valid `config.yml` file in a `.circleci` folder at the root of your repository. In this example, the `config.yml` contains the following:

>     ```sh

Reviewer 3: This should be yaml.

Author: fixed

          }
        }
      ]
    }
    ```
    
That's great! Hopefully everything is working for you up to this point. Let's move on to performing something that might be a bit more useful.

5. One of the benefits of the CircleCI API v2 is the ability to remotely trigger pipelines with parameters. The following code snippet simply triggers a pipeline via `curl` without any body parameters:

    ```sh
    curl -X POST https://circleci.com/api/v2/project/gh/{YOUR_USER_NAME}/hello-world/pipeline \
      -H 'Content-Type: application/json' \
      -H 'Accept: application/json' \
      -H "Circle-Token: $CIRCLECI_TOKEN" \
      
    # Which returns:
    {
      "number": 2,
      "state": "pending",
      "id": "e411ea74-c64a-4d60-9292-115e782802ed",
      "created_at": "2020-01-15T15:32:36.605Z"
    }
    ```

While this alone can be useful, we want to be able to customize parameters of the pipeline when we send this POST request. By including a body parameter in the `curl` request (via the `-d` flag), we can customize specific attributes of the pipeline when it runs: pipeline parameters, the branch, or the git tag. Below, we are telling the pipelines to trigger for "my-branch"

>     ```sh

Reviewer 3: This code block is broken, the backticks are showing. I think it shouldn't be indented?

Author: fixed

        type: string
        default: "latest"
    
    jobs:
      build:
        docker:
          - image: circleci/node:<< pipeline.parameters.image-tag >>
        environment:
          IMAGETAG: << pipeline.parameters.image-tag >>
        steps:
          - run: echo "Image tag used was ${IMAGETAG}"
    ```

You will need to declare the parameters you expect to receive from the API. In this case, under the `parameters` key, we definte an "image-tag" to be expected in the JSON payload of a POST request to the _Trigger New Pipeline_ endpoint.

7. Now we can run a `curl` request that passes variables in a POST request, similar to the following:

    ```sh
    curl -u ${CIRCLECI_TOKEN}: -X POST --header "Content-Type: application/json" -d '{
      "parameters": {
        "image-tag": "4.8.2"
      }
    }' https://circleci.com/api/v2/project/gh/{YOUR_USER_NAME}/hello-world/pipeline
    ```

This concludes the end-to-end example of using the V2 API. For more detailed information about other endpoints you may wish to call, please refer to the [CircleCI API v2 Documentation]({{site.baseurl}}/api/v2/#circleci-api) for an overview of all endpoints currently available.

> # Example API Use Cases

Reviewer 3: Arguably the section above was already an example use case.

Author: fixed

```json
  {
	"web_url": "string",
	"project": {
		"slug": "gh/CircleCI-Public/api-preview-docs",
		"name": "api-preview-docs",
		"external_url": "https://github.com/CircleCI-Public/api-preview-docs"
	},
	"parallel_runs": [{
		"index": 0,
		"status": "string"
	}],
	"started_at": "2020-01-24T11:33:40Z",
	"latest_workflow": {
		"id": "string",
		"name": "build-and-test"
	},
	"name": "string",
	"executor": {
		"type": "string",
		"resource_class": "string"
	},
	"parallelism": 0,
	"status": null,
	"number": 0,
	"pipeline": {
		"id": "string"
	},
	"duration": 0,
	"created_at": "2020-01-13T18:51:40Z",
	"messages": [{
		"type": "string",
		"message": "string",
		"reason": "string"
	}],
	"contexts": [{
		"name": "string"
	}],
	"organization": {
		"name": "string"
	},
	"queued_at": "2020-01-13T18:51:40Z",
	"stopped_at": "2020-01-13T18:51:40Z"
}
```
Notice in the example above that you will receive very specific information about your job, including specific project and workflow details for the job, the date and time the job started and then finished, and job-specific information such as the executor type used, current status of the job, and the duration of the job.

> For a more detailed breakdown of each value returned in this request, please refer to the [Get Job Details](https://circleci.com/docs/api/v2/#get-job-details)section of the *CircleCI API v2 Reference Guide*.

Reviewer 3: There's a space missing before `section`.

Author: fixed

---
layout: classic-docs
title: "CircleCI API Developer's Guide"
short-title: "Developer's Guide"
description: "API cookbook for internal and external CircleCI developers"
categories: [getting-started]
order: 1
---

This *API Developer's Guide* was written to assist developers in quickly and easily making API calls to CircleCI services to return detailed information about users, pipelines, projects and workflows.

* TOC
{:toc}

# API Overview

> The CircleCI platform provides a powerful API that enables users to retrieve detailed information about users, jobs, workflows and pipelines.

Reviewer 4: ~\~ Is there a way to rephrase this without including the word "powerful"? The API can also be used to interact with CircleCI, e.g. triggering pipelines, updating settings etc.

Author: fixed in document

## Authentication and Authorization

The CircleCI API utilizes token-based authentication to manage access to the API server and validate that a user has permission to make API requests. Before you can make an API request, you must first add an API token and then verify that you are authenticated by the API server to make requests. The process to add an API token and have the API server authenticate you is described in the sections below.

**Note** You may use the API token as the username for HTTP Basic Authentication, by passing the `-u` flag to the `curl` command.

### Add an API Token

To add an API token, perform the steps listed below.

1.  Add an API token from the [Personal API Tokens page](https://account.circleci.com/tokens). 
2.  To test your token call the API using the command below. You will need to set your API token as an environment variable before making a cURL call.

    ```sh
    export CIRCLE_TOKEN=<your_token>

>     curl https://circleci.com/api/v2/me -H "Circle-Token: $CIRCLE_TOKEN"

Reviewer 4: ~ suggest using long-form args, e.g. `--header "Circle-Token: $CIRCLE_TOKEN"`

Author: fixed throughout document

<!---
## API Syntax

When making an API request, make sure you follow standard REST API syntax and formatting. Adhering to proper REST API syntax ensures the API server can properly process your request and return a valid JSON response. To make a request to the CircleCI API, use the following format:

`https://circleci.com/api/v2`

Where:

- `https://circleci.com/api/` is the resource URL for the API being called.
- `api` is the service being called.
- `v2` is the API version.
-->

# Getting Started with the API

The CircleCI API v2 shares similarities with previous API versions in that it identifies your projects using repository name. For instance, if you want to pull information from CircleCI about the GitHub repository https://github.com/CircleCI-Public/circleci-cli you can refer to that in the CircleCI API as `gh/CircleCI-Public/circleci-cli`, which is a “triplet” of the project type (or, VCS provider), the name of your “organization”, and the name of the repository. 

> For the project type you can use `github` or `bitbucket` as well as the shorter forms `gh` or `bb`, which are supported in API v2. The `organization` is your username or organization name in your version control system.

Reviewer 4: ```suggestion
For the project type you can use `github` or `bitbucket`. API v2 also supports the shorter forms `gh` or `bb`. The `organization` is your username or organization name in your version control system.
```

Author: Fixed

* [Download artifacts](#download-artifacts)

## Get Project Details

You may often find that it would be helpful to retrieve information about a specific project, including the name of the organization the project belongs to, the version control system (vcs) that hosts the project, and other details. The CircleCI API enables you to return this and other information by making a single GET request to the `project/{project-slug}` endpoint by passing the `project-slug` parameter.

**Note** If you would like more detailed information about a project, or simply need a refresher on the specifics of a project, please refer to the CircleCI [Projects](https://circleci.com/docs/2.0/projects/) page.

### Steps

Of the several project-related API endpoints available with CircleCI API v2, making a GET request to the `/project/{project-slug}`endpoint enables you to return detailed information about a specific project by passing the `project_slug` parameter with your request.

**Note** In this example, please note that whenever you see curly brackets `{}`, this represents a variable that you must manually enter in the request.

To return project details, perform the following steps:

> 1. Declare the parameters you expect to receive from the API. For this GET API call, under the `parameters` key, define the `project_slug` parameter you want returned in the JSON payload in your `curl` request as follows:

Reviewer 2: I can't follow what the test in this step is asking me to do - sounds like I have to define project_slug somehow but then the example doesn't show how? - "...under the `parameters key`..."

Author: added more clarification to how to use project slug.

Reviewer 2: What I meant was I don't understand what is meant by "under the `parameters` key, define the `project_slug` parameter you want returned..." ... I mean I get how to define the `project_slug` but the step still isn't clear to me, (e.g what's the `parameters` key?) We could explicitly state the elements as Tyler did in his example, and use a table to show what they all are like I added in the download artifacts use case (as you suggested)? Might be clearer for the reader.

---
layout: classic-docs
title: "CircleCI API Developer's Guide"
short-title: "Developer's Guide"
description: "API cookbook for internal and external CircleCI developers"
categories: [getting-started]
order: 1
---

This *API Developer's Guide* was written to assist developers in quickly and easily making API calls to CircleCI services to return detailed information about users, pipelines, projects and workflows.

* TOC
{:toc}

# API Overview

The CircleCI platform provides an API that enables users to interact with the CircleCI platform (e.g. triggering pipelines, updating settings etc, in addition to retrieving detailed information about users, jobs, workflows and pipelines.

## API Classes

The current categories of API v2 endpoints are:

* Authentication
* Pipeline

> * Workflows

Reviewer 2: I think insights (preview) needs adding here?

Author: @rosieyohannan I intentionally left this out of the list because we are not exposing anything related to insights in this initial release.

Reviewer 2: Ah ok, I only ask because I saw it's there in the public API docs

### User Endpoints

| Endpoint               | Description                       
|------------------------|-------------------------------------------------------
| GET /user/{id}         | This endpoint enables you to return information about a specific user by making a GET request and passing the `id` parameter in the request |
| GET /me/collaborations | This endoint enables you to return a list of organizations that the user is a member of, or is a collaborator by making a GET request and passing the `collaborations` parameter in the request. |
{: class="table table-striped"}

### Pipeline API Requests

| Endpoint                                  | Description                                                         | 
|-------------------------------------------|---------------------------------------------------------------------|
| GET /pipeline/{pipeline-id}               | Returns an individual pipeline.                                     |
| GET /pipeline/{pipeline-id}/config        | Returns the configuration of a specific pipeline.                   |

> | GET /pipeline/{pipeline-id}/workflow      | Returns a list of workflows for a specific pipeline.                |

Reviewer 2: @thomasr8 I think `/project/{project_slug}` needs adding to the start of these to be consistent

Reviewer 2: Also - from what I can tell there is a pipeline numbers, and a pipeline ID. I think these endpoints should use {pipeline-number}? Not sure what the difference is though!

Author: Fixed

This section provides detailed information on how you can perform the following tasks and operations:

* [Get project details](#get-project-details)
* [Get job details](#get-job-details)
* [Download artifacts](#download-artifacts)

## Get Project Details

You may often find that it would be helpful to retrieve information about a specific project, including the name of the organization the project belongs to, the version control system (vcs) that hosts the project, and other details. The CircleCI API enables you to return this and other information by making a single GET request to the `project/{project-slug}` endpoint by passing the `project-slug` parameter.

**Note** If you would like more detailed information about a project, or simply need a refresher on the specifics of a project, please refer to the CircleCI [Projects](https://circleci.com/docs/2.0/projects/) page.

> ### Prerequisites

Reviewer 5: This section is repeated several times per use-case. I think we can move it up into a top level section, or just mention it once and say it applied to all "Additional API Use Cases"

Author: fixed in document

		"index": 0,
		"status": "string"
	}],
	"started_at": "2020-01-24T11:33:40Z",
	"latest_workflow": {
		"id": "string",
		"name": "build-and-test"
	},
	"name": "string",
	"executor": {
		"type": "string",
		"resource_class": "string"
	},
	"parallelism": 0,
	"status": null,
	"number": 0,
	"pipeline": {
		"id": "string"
	},
	"duration": 0,
	"created_at": "2020-01-13T18:51:40Z",
	"messages": [{
		"type": "string",
		"message": "string",
		"reason": "string"
	}],
	"contexts": [{
		"name": "string"
	}],
	"organization": {
		"name": "string"
	},
	"queued_at": "2020-01-13T18:51:40Z",
	"stopped_at": "2020-01-13T18:51:40Z"
}
```
Notice in the example above that you will receive very specific information about your job, including specific project and workflow details for the job, the date and time the job started and then finished, and job-specific information such as the executor type used, current status of the job, and the duration of the job.

For a more detailed breakdown of each value returned in this request, please refer to the [Get Job Details](https://circleci.com/docs/api/v2/#get-job-details) section of the *CircleCI API v2 Reference Guide*.

## Download Artifacts

> The following section details the steps you need to follow to download artifacts generated when a job is run. We will go through the steps of returning a list of artifacts for a job, and then changing the command we run to download the full set of artifacts. If you are looking for instructions to download the latest artifacts for a pipeline, without needing to specify a job number, see our [API v1.1 guide](https://circleci.com/docs/2.0/artifacts/#downloading-all-artifacts-for-a-build-on-circleci) – keep checking back here as this functionality will be added to API v2 in the future.

Author: ```suggestion
If you are looking for instructions on how to download the latest artifacts for a pipeline, without needing to specify a job number, see our [API v1.1 guide](https://circleci.com/docs/2.0/artifacts/#downloading-all-artifacts-for-a-build-on-circleci) – keep checking back here as this functionality will be added to API v2 in the future.
```

### Prerequisites

Before making an API call, make sure you have met the following prerequisites:

* You have set up a GitHub or BitBucket account with a repository to use with CircleCI.
* You have completed CircleCI onboarding and you have a project setup.
* You have a [personal API token](https://circleci.com/account/api).
* You have been [authenticated](#get-authenticated) to make API calls to the server.

### Steps

> 1. First we will ensure your api token is set as an environment variable. You maybe have already done this during authentication, but if not, run the following command in your terminal, substituting your personal api token:

Author: ```suggestion
1. First, we will ensure your API token is set as an environment variable. You maybe have already done this during authentication, but if not, run the following command in your terminal, substituting your personal API token:
```

Reviewer 2: fixed

> 1. First we will ensure your api token is set as an environment variable. You maybe have already done this during authentication, but if not, run the following command in your terminal, substituting your personal api token:

Author: ```suggestion
1. First, ensure your API token is set as an environment variable. You maybe have already done this during authentication, but if not, run the following command in your terminal, substituting your personal API token:
```

    ```
    export CIRCLECI_TOKEN={your_api_token}
    ```

> 2.  Next we need the job number for the job you want to get artifacts for. You can find job numbers in the UI - either in the breadcrumbs on the Job Details page, or in the URL.

Author: ```suggestion
2.  Next, you need to retrieve the job number for the job you want to get artifacts for. You can find job numbers in the UI - either in the breadcrumbs on the Job Details page, or in the URL.
```

    INSERT SCREENSHOT HIGHLIGHTING JOB NUMBER

> 3.  Next we can use `curl` to return a list of artifacts for a specific job. Note, you can format JSON responses by piping the `curl` command into the `jq` utility if you have it installed: `curl ... | jq`:

Author: ```suggestion
3.  Next, use the `curl` command to return a list of artifacts for a specific job. Note, you can format JSON responses by piping the `curl` command into the `jq` utility if you have it installed: `curl ... | jq`:
```

    ```sh
    curl -X GET https://circleci.com/api/v2/project/{VCS}/{org-name}/{repo-name}/{job-number}/artifacts \
    -H 'Content-Type: application/json' \
    -H 'Accept: application/json' \
    -H "Circle-Token: $CIRCLECI_TOKEN" | jq\
    ```

>     Check the table below for help with formatting your API call correctly.

Author: @rosieyohannan I really like how you have broken down the parameters here for the API call. Tyler and I did not do this for our API use cases, so do you think we should make this change as well? I am just thinking about consistency in the document.

    | Parameter    | Description                                                             |
    | ------------ | ----------------------------------------------------------------------- |
    | `VCS`        | Version Control System - either `github`/`gh` or `bitbucket`/`bb`       |
    | `org-name`   | Organization name, or your personal username to your VCS                |
    | `repo-name`  | Name of your project repo                                               |
    | `job-number` | The number for the job you want to download artifacts from - see step 2 |
    {: class="table table-striped"}
    

>     You should get a list of artifacts back - if the job you chose has artifacts associated with it. Here's and extract from the output when requesting artifacts for a job that builds these docs:

Author: ```suggestion
    You should get a list of artifacts back - if the job you selected has artifacts associated with it. Here's an extract from the output when requesting artifacts for a job that builds these docs:
```

Author: ```suggestion
    You should receive a list of artifacts back - if the job you chose has artifacts associated with it. Here's an extract from the output when requesting artifacts for a job that builds these docs:
```

    ```
    {
      "path": "circleci-docs/assets/img/docs/walkthrough6.png",
      "node_index": 0,
      "url": "https://53936-48750547-gh.circle-artifacts.com/0/circleci-docs/assets/img/docs/walkthrough6.png"
    },
    {
      "path": "circleci-docs/assets/img/docs/walkthrough7.png",
      "node_index": 0,
      "url": "https://53936-48750547-gh.circle-artifacts.com/0/circleci-docs/assets/img/docs/walkthrough7.png"
    },
    {
      "path": "circleci-docs/assets/img/docs/walkthrough8.png",
      "node_index": 0,
      "url": "https://53936-48750547-gh.circle-artifacts.com/0/circleci-docs/assets/img/docs/walkthrough8.png"
    },
    ```

    CHECK ITS OK TO INCLUDE THIS

> 4. Next we can extend this API call to download the artifacts. Navigate to the location you would like to download the artifacts to, and run the following, remembering to substitute your values:

Author: ```suggestion
4. Next, you may extend this API call to download the artifacts. Navigate to the location you would like to download the artifacts to, and run the following command, remembering to substitute your own values in the request:
```

     ```sh
    curl -X GET https://circleci.com/api/v2/project/{VCS}/{org-name}/{repo-name}/{job-number}/artifacts \
    -H 'Content-Type: application/json' \
    -H 'Accept: application/json' \
    -H "Circle-Token: $CIRCLECI_TOKEN" \
    | grep -o 'https://[^"]*' \
    | wget -v -i -
    ```

>   `grep` is used to locate all the URLs for downloading the job artifacts, and `wget` is used to perform the download.

Author: ```suggestion
  `Note: `grep` is used to locate all the URLs for downloading the job artifacts, while `wget` is used to perform the download.
```

     ```sh
    curl -X GET https://circleci.com/api/v2/project/{VCS}/{org-name}/{repo-name}/{job-number}/artifacts \
    -H 'Content-Type: application/json' \
    -H 'Accept: application/json' \
    -H "Circle-Token: $CIRCLECI_TOKEN" \
    | grep -o 'https://[^"]*' \
    | wget -v -i -
    ```

    `Note: `grep` is used to locate all the URLs for downloading the job artifacts, while `wget` is used to perform the download.

# Reference

## API Endpoints

The CircleCI v2 API, and its associated endpoints allow you to make HTTP calls to designated endpoints developed in the underlying CircleCI API architecture. These endpoints provide programmatic access to CircleCI services, including pipelines, workflows, and jobs. 

Before working with the API and making calls, you should first have an understanding of the various API requests you can make to API v2 endpoints. The sections below describe these API requests.

### User Endpoints

| Endpoint               | Description                       
|------------------------|-------------------------------------------------------

> | GET /user/{id}         | This endpoint enables you to return information about a specific user by making a GET request and passing the `id` parameter in the request |

Reviewer 2: @thomasr8 I stuck `GET` in for both these user endpoints as it wasn't there before and seemed inconsistent with the other tables. Just wanted to check with you if it's actually correct :-)

Author: @rosieyohannan Yes, that is correct! Good catch 👍

---
layout: classic-docs
title: "CircleCI API Developer's Guide"
short-title: "Developer's Guide"
description: "API cookbook for internal and external CircleCI developers"
categories: [getting-started]
order: 1
---

This *API Developer's Guide* was written to assist developers in quickly and easily making API calls to CircleCI services to return detailed information about users, pipelines, projects and workflows.

* TOC
{:toc}

# API Overview

The CircleCI platform provides a powerful API that enables users to retrieve detailed information about users, jobs, workflows and pipelines.

## API Classes

> The current classes of the API v2 endpoints are:

Reviewer 6: Not sure how common it is to refer to different endpoints as "classes"? 

> The current available API v2 endpoints are:

Author: replaced "classes" with "categories"

* Authentication
* Pipeline
* Workflows
* User (Preview)
* Project (Preview)
* Job (Preview)

> 

Reviewer 6: Should we add the new Insights API in preview?
https://circleci.com/api/v2/insights
CC @KunalJain

Author: @KyleTryon We will add Insights API endpoints in Q1.

**Note:** Portions of the CircleCI API v2 remain under “Preview”. Preview endpoints are not yet fully supported or considered generally available. Breaking changes to API v2 Preview endpoints are planned in advance and are announced in the API v2 breaking changes log.

## Authentication and Authorization

The CircleCI API utilizes token-based authentication to manage access to the API server and validate that a user has permission to make API requests. Before you can make an API request, you must first add an API token and then verify that you are authenticated by the API server to make requests. The process to add an API token and have the API server authenticate you is described in the sections below.

**Note** You may use the API token as the username for HTTP Basic Authentication, by passing the `-u` flag to the `curl` command.

### Add an API Token

To add an API token, perform the steps listed below.

> 1.  Add an API token from your [account dashboard](https://circleci.com/account/api).

Reviewer 5: Fixed

> 2.  To test it, [view it in your browser](https://circleci.com/api/v2/me) or call the API using the command below.

Reviewer 5: Fixed!

    ```sh

>     $ curl https://circleci.com/api/v2/me

Reviewer 5: Fixed!

    ```

3.  You should see a JSON response similar to the example shown below.

    ```json
    {
      "user_key_fingerprint" : null,
      "days_left_in_trial" : -238,
      "plan" : "p16",
      "trial_end" : "2011-12-28T22:02:15Z",
      "basic_email_prefs" : "smart",
      "admin" : true,
      "login" : "someuser"
    }
    ```

**Note:** All API calls are made in the same way, by making standard HTTP calls, using JSON, a content-type, and your API token. Please note that the JSON examples shown in this document are not comprehensive and may contain additional JSON response fields not shown in the example, based on user input and fields.

### Get Authenticated

To be authenticated by the API server, add an API token using your [account dashboard](https://circleci.com/account/api). To use the API token, add it to the `circle-token` query param:

```sh

> curl "https://circleci.com/api/v1.1/me?circle-token=:token"

Reviewer 5: I'm not sure why this section is in here :

<details> Excerpt

### Get Authenticated

To be authenticated by the API server, add an API token using your [account dashboard](https://circleci.com/account/api). To use the API token, add it to the `circle-token` query param:

```sh
curl "https://circleci.com/api/v1.1/me?circle-token=:token"
```

Alternatively, you can use the API token as the username for HTTP Basic Authentication, by passing the `-u` flag to the `curl` command:

```sh 
curl -u <circle-token>: "https://circleci.com/api/..."
```

**Note:** The colon ":" notifies `curl` that there is no password being passed.

</details>

The section `### Get Authenticated` says mostly the same thing as `### Add an API token`, and it refers to API 1.1 (both the url, the use of `?circle-token=:token` when we should be using the `-H` token method.  I think we could just remove this section entirely — @thomasr8 what do you think?

> curl "https://circleci.com/api/v1.1/me?circle-token=:token"

Author: @teesloane You are right. This is redundant information. I went ahead and removed the "Get Authenticated" section from the page.

```

Alternatively, you can use the API token as the username for HTTP Basic Authentication, by passing the `-u` flag to the `curl` command:

```sh 
curl -u <circle-token>: "https://circleci.com/api/..."
```

**Note:** The colon ":" notifies `curl` that there is no password being passed.

<!---
## API Syntax

When making an API request, make sure you follow standard REST API syntax and formatting. Adhering to proper REST API syntax ensures the API server can properly process your request and return a valid JSON response. To make a request to the CircleCI API, use the following format:

`https://circleci.com/api/v2`

Where:

> - `https://circleci.com` is the resource URL for the API being called.

Reviewer 5: I guess this is commented out, but I will update this nonetheless :) - not sure if it will get removed eventually anyway.

- `api` is the service being called.
- `v2` is the API version.
-->

# Getting Started with the API

> The CircleCI API v2 is backwards-compatible with previous API versions in the way it identifies your projects using repository name. For instance, if you want to pull information from CircleCI about the GitHub repository https://github.com/CircleCI-Public/circleci-cli you can refer to that in the CircleCI API as `gh/CircleCI-Public/circleci-cli`, which is a “triplet” of the project type, the name of your “organization”, and the name of the repository. For the project type you can use `github` or `bitbucket` as well as the shorter forms `gh` or `bb`, which are supported in API v2. The `organization` is your username or organization name in your version control system.

Reviewer 5: Resolved!

If you receive a 400 HTTP status code, there is a problem with the request and the server is unable to successfully process the request. The following status codes may be returned with your request:

* 401 - Unauthorized
* 403 - Forbidden
* 404 - Not Found

## 500 Status Code

If you receive a 500 HTTP status code, there is a problem with the server and the request cannot be processed. If you encounter a 500 response, the error will be logged and CircleCI will work to resolve the error. The following 500 HTTP status codes could potentially be returned with your request:

* 500 - Internal Server Error

# REST API Tutorial

For more detailed information about HTTP status codes, refer to the [rest api tutorial](http://www.restapitutorial.com/httpstatuscodes.html).

# Example End-to-End API Request

The following section details the steps you would need, from start to finish, to make an API call. This section includes creating a "demo repository" called "hello-world", however, you can use a pre-existing repository to follow along if you choose.

## Prerequisites

* A GitHub or BitBucket account with a repository to setup with CircleCI.
* Completion of the CircleCI onboarding.

> * **Optional**: You can format JSON responses by piping the `curl` command into the `jq` utility if you have it installed: `curl ... | jq`.

Reviewer 6: I would remove the portion mentioning JQ. That's a little removed from the API.

> * **Optional**: You can format JSON responses by piping the `curl` command into the `jq` utility if you have it installed: `curl ... | jq`.

Reviewer 5: This makes sense. I have removed JQ.

## Steps

1. On your VCS provider, create a repository. The repo for this example will be called `hello-world`.

2. Next, follow the onboarding for a new project on CircleCI. You can access onboarding by visiting the application and clicking on "Add Projects" in the sidebar or by going to the link: https://onboarding.circleci.com/project-dashboard/{VCS}/{ORG_NAME} where `VCS` is either `github` (or `gh`) or `bitbucket` (or `bb`) and `ORG_NAME` is your organization or personal VCS username. Find your project in the onboarding list and click `Setup Project`. After completing an onboarding, you should have a valid `config.yml` file in a `.circleci` folder at the root of your repository. In this example, the `config.yml` contains the following:

    ```yaml
    # Use the latest 2.1 version of CircleCI pipeline process engine. See: https://circleci.com/docs/2.0/configuration-reference
    version: 2.1
    # Use a package of configuration called an orb.
    orbs:
    # Declare a dependency on the welcome-orb
      welcome: circleci/welcome-orb@0.4.1
    # Orchestrate or schedule a set of jobs

>       workflows:

Reviewer 5: Fixed!

      # Name the workflow "welcome"
      welcome:
      # Run the welcome/run job in its own container
        jobs:
        - welcome/run
    ```

3. Add an API token from your [account dashboard](https://circleci.com/account/api). Be sure to write down and store your API token in a secure place once you generate it.

4. It's time to test out your API token using `curl` to make sure everything works. The following code snippets demonstrate querying all pipelines on a project. Please note that in the example below, the values within curly braces (`{}`) need to be replaced with values specific to your username/orgname.

    ```sh
    # First: set your CircleCI token as an environment variable
    export CIRCLECI_TOKEN={your_api_token}
        
    curl --header "Circle-Token: $CIRCLECI_TOKEN" \
      --header 'Accept: application/json'    \
      --header 'Content-Type: application/json' \

>       https://circleci.com/api/v2/project/gh/{USER_NAME}/hello-world/pipeline 

Reviewer 6: Replace 'gh' with {VCS} or {VCS-Provider}. 

All of these should be valid in 2.0:
gh
github
bb
bitbucket

Reviewer 5: Fixed!

    ```

    You will likely receive a long string of unformatted JSON. After formatting, it should look like so:

>     ```sh

Reviewer 6: sh > json

      "created_at": "2020-01-15T15:32:36.605Z"
    }
    ```

    While this alone can be useful, we want to be able to customize parameters of the pipeline when we send this POST request. By including a body parameter in the `curl` request (via the `-d` flag), we can customize specific attributes of the pipeline when it runs: pipeline parameters, the branch, or the git tag. Below, we are telling the pipelines to trigger for "my-branch"

    ```sh
    curl -X POST https://circleci.com/api/v2/project/gh/teesloane/hello-world/pipeline \
    -H 'Content-Type: application/json' \
    -H 'Accept: application/json' \
    -H "Circle-Token: $CIRCLE_TOKEN" \
    -d '{ "branch": "bar" }' 
    ```

6. Let's move on to a more complex example: triggering a pipeline and passing a parameter that can be dynamically substituted into your configuration. In this example, we will pass a docker image tag to our docker-executor key. First, we will need to modify the `.circleci/config.yml` to be a little more complex than the standard "Hello World" sample provided by the onboarding.

    ```yaml
    version: 2.1
    jobs: 
      build: 
        docker: 
          - 

>             image: "circleci/node:<< pipeline.parameters.image-tag >>"

Reviewer 7: This should be on the same line as the dash.

Author: @teesloane fixed this for us. Thanks @teesloane

        environment: 
          IMAGETAG: "<< pipeline.parameters.image-tag >>"
        steps: 
          - 

>             run: "echo \"Image tag used was ${IMAGETAG}\""

Reviewer 7: This should be on the same line as the dash.

Also, instead of escaping quotes, you can do this:

```
- run: "echo 'Image tag used was ${IMAGETAG}'"
```

or not use them on the outside at all:

```
- run: echo "Image tag used was ${IMAGETAG}"
```

Reviewer 5: Fixed! (and thanks for the tips!)

    parameters: 
      image-tag: 
        default: latest
        type: string
    ```

    You will need to declare the parameters you expect to receive from the API. In this case, under the `parameters` key, we definte an "image-tag" to be expected in the JSON payload of a POST request to the _Trigger New Pipeline_ endpoint.

7. Now we can run a `curl` request that passes variables in a POST request, similar to the following:

    ```sh
    curl -u ${CIRCLECI_TOKEN}: -X POST --header "Content-Type: application/json" -d '{
      "parameters": {
        "image-tag": "4.8.2"
      }

>     }' https://circleci.com/api/v2/project/gh/{YOUR_USER_NAME}/hello-world/pipeline

Reviewer 6: replace gh here as well
